ambry-replication/src/main/java/com.github.ambry.replication/ReplicaThread.java

```diff
@@ -485,14 +526,28 @@ private ReplicaMetadataResponse getReplicaMetadataResponse(List<RemoteReplicaInf
           remoteReplicaInfo.getReplicaId(), messageInfo.getStoreKey());
     }
 
-    Set<StoreKey> missingStoreKeys = remoteReplicaInfo.getLocalStore().findMissingKeys(storeKeysToCheck);
-    for (StoreKey storeKey : missingStoreKeys) {
+    Map<StoreKey, StoreKey> unconvertedToConverted = new HashMap<>();
+    for (StoreKey storeKey : storeKeysToCheck) {
+      unconvertedToConverted.put(storeKey, storeKeyConverter.getConverted(storeKey));
+    }
+
+    twoWayConverterCache.setRemoteToLocal(unconvertedToConverted);
+    List<StoreKey> convertedKeys = new ArrayList<>();
+    for (StoreKey val : unconvertedToConverted.values()) {
+      if (val != null) {
+        convertedKeys.add(val);
+      }
+    }
+    Set<StoreKey> convertedMissingStoreKeys = remoteReplicaInfo.getLocalStore().findMissingKeys(convertedKeys);
+    Set<StoreKey> unconvertedStoreKeys = new HashSet<>();
+    for (StoreKey storeKey : convertedMissingStoreKeys) {
       logger.trace("Remote node: {} Thread name: {} Remote replica: {} Key missing id: {}", remoteNode, threadName,
           remoteReplicaInfo.getReplicaId(), storeKey);
+      unconvertedStoreKeys.add(twoWayConverterCache.getRemote(storeKey));
```

Comment:
Is this the only reason to have a 2-way cache? If so, I think you can avoid the 2-way thing. Instead, simply keep the `originalToConverted` map (that is returned by the batch call) and do the following here instead:
```java
for (Map.Entry<StoreKey, StoreKey> entry: originalToConverted.entrySet()) {
  if (convertedMissingStoreKeys.contains(entry.getValue()) {
    missingKeySet.add(entry.getKey());
  }
}
```

Reply:
Tested and it works.  I'll get rid of TwoWayConverter